gpcontrib/arenadata_toolkit/arenadata_toolkit.c

```diff
@@ -1,4 +1,191 @@
+#include <dirent.h>
+#include <sys/stat.h>
+#include <ctype.h>
+
 #include "postgres.h"
 #include "fmgr.h"
+#include "funcapi.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "nodes/execnodes.h"
+#include "cdb/cdbvars.h"
+#include "utils/builtins.h"
+#include "utils/relfilenodemap.h"
+#include "utils/timestamp.h"
+#include "catalog/pg_tablespace.h"
 
 PG_MODULE_MAGIC;
+
+typedef struct
+{
+	char     *datpath;
+	DIR      *dirdesc;
+	TupleDesc tupdesc;
+} user_fctx_data;
+
+static bool should_skip_file(const char *filename)
+{
```

Comment:
What do you think about rewrite this function like code below:
```c
typedef struct pattern {
	char *str;
	bool prefix;
} pattern;

static bool should_skip_file(char *fname) {
	// this part of code has a problem with a static initialization
    // maybe make patterns array insted of array to pattern pointers and rewrite loop as for (int i = 0; i < sizeof(patterns)/sizeof(pattern); i++)
    pattern *patterns[] = {
		&(pattern){"pg", 1}, &(pattern){"t_", 1},
		&(pattern){"_vm", 0}, &(pattern){"_fsm", 0}, &(pattern){"_init", 0}, NULL
	};

	int len = strlen(fname);
	for (pattern **p = patterns; *p != NULL; p++) {
		pattern *pat = *p;
		int plen = strlen(pat->str);
		if (len - plen < 0)
			continue;

		int startPos = pat->prefix ? 0 : len - plen;
		if (pg_strncasecmp(fname + startPos, pat->str, plen) == 0)
			return true;
	}
	return false; 
}
```

Reply:
I suggested rewriting this function in such a way as to avoid calls to `pg_strncasecmp`, because all patterns are known to us in advance and they are simple.